packages/sdk-middleware-http/src/http.js

```diff
@@ -128,6 +128,21 @@ export default function createHttpMiddleware({
             })
             return
           }
+          if (res.status === 503 && enableRetry)
```

Comment:
I don't think that retrying on a `500` makes a lot of sense, because it could mean that the problem is persistent and can't be resolved with a retry. But doing a retry on a response with a `502` or `504` status could actually make sense.

Reply:
There are many reasons why one could hit 500 error. You might hit smaller node which might have not enough memory and return 500 and on retry you might be hitting different/bigger node. I would not expect API to throw 500 errors repeatedly and before i fail whole process i would rather try to recover. I thought of 502 and 504 errors (on which btw. we retry in older sphere-node-sdk automatically already) but since API Extension has been introduced those codes rather remind BadRequest so generic retry would not be reasonable anymore: https://docs.commercetools.com/http-api-errors.html#extension-error-responses

Reply:
Exactly this is the problem why you should not retry on 500. It's unknown to tell what the reason for the 500 is. It could be that there are no resources left on the node, some application error or whatever. In case you retry you would leverage the problem by possibly overloading other nodes.

Only 503 has the clear semantic that the application itself is working but overloaded for the moment.

Reply:
Btw the 504 should also not be retried when not sure this comes from an extension as the initial request has been accepted but the answer could not provided timely. The backend may still process the initial request and by retrying this would either lead to duplicates or in case of updates may be revoked because of mismatching versions.

And retrying on 502 is useless cause the recovery of connectivity issues most likely will take longer than the retry timeouts.

Reply:
We use exponential backoff on retry so i do not know why do you think it would put more pressure? Having rather failed customer process than hypothetically overloaded API does not sound right too. It is known pattern to retry on 500 errors: https://cloud.google.com/apis/design/errors
https://docs.aws.amazon.com/general/latest/gr/api-retries.html

Reply:
Okay. Still when looking at it than from a semantic view 500 is unspecified. Most likely this happens in case of an application error which is not properly handled. Retrying makes no sense cause the application error will not be gone in the retry timouts.

Reply:
> would either lead to duplicates or in case of updates may be revoked because of mismatching versions

No duplicates as everywhere you either have unique key/slug/sku/whatever/version in play. I would rather have after first retry logged BadRequest and correct data in system than logged 502 and eventually no data written.

> And retrying on 502 is useless cause the recovery of connectivity issues most likely will take longer than the retry timeouts

Even if it is "Most likely" i would rather retry as you can influence waiting time greatly by modifying both retryCount and delay. Very useful especially for async processes as you need to succeed at some point anyway and giving up is not an option :)

Reply:
We are still talking about a generic http client and not architecture of the API do we? In the actual stack it just makes no sense. And the PR adds a feature which was described to do a retry on 503 but this was not working.

Reply:
I never chaged the topic. Just gave you already couple of examples why retry on 500 makes sense. I just do not agree that it should retry on 503 only but still glad for the fix.

Reply:
I would not encourage a retry on a 500, since this is a response with no actual _context_* and the reason for it could be __anything__. even if we were to continue with 500 + retry, we're making harder for ourselves in the API to detect a potential problem e.g via metrics

_context_: any information that could help the client determine how it should react once accepted.